src/injected/visualization/table-headers-formatter.ts

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { DialogRenderer } from '../dialog-renderer';
+import { AssessmentVisualizationInstance } from '../frameCommunicators/html-element-axe-results-helper';
+import { FailureInstanceFormatter } from './failure-instance-formatter';
+import { DrawerConfiguration } from './formatter';
+
+export interface HeadingStyleConfiguration {
+    borderColor: string;
+    fontColor: string;
+}
+
+export interface StyleComputer {
+    getComputedStyle(elt: Element, pseudoElt?: string): CSSStyleDeclaration;
+}
+
+export class TableHeadersAttributeFormatter extends FailureInstanceFormatter {
+    private readonly cellColor = '#6600CC';
+    private readonly headerColor = '#0066CC';
+
+    constructor() {
+        super();
+    }
+
+    public getDialogRenderer(): DialogRenderer {
+        return null;
+    }
+
+    public getDrawerConfiguration(
+        element: HTMLElement,
+        data: AssessmentVisualizationInstance,
+    ): DrawerConfiguration {
+        const isHeader = this.isHeader(element);
+
+        const style = {
+            borderColor: isHeader ? this.headerColor : this.cellColor,
+            fontColor: '#FFFFFF',
+        };
+
+        const text = isHeader ? this.getTextForHeader(element) : this.getTextForCell(element);
+
+        return {
+            textBoxConfig: {
+                fontColor: style.fontColor,
+                text,
+                background: style.borderColor,
+            },
+            borderColor: style.borderColor,
+            outlineStyle: 'solid',
+            showVisualization: true,
+            textAlign: 'right',
+            failureBoxConfig: this.getFailureBoxConfig(data),
+        };
+    }
+
+    private isHeader(element: HTMLElement): boolean {
+        if (element.matches('th')) {
+            return true;
+        }
+        const role = this.getAttribute(element, 'role');
+        if (role === 'columnheader' || role === 'rowheader') {
+            return true;
+        }
+        return false;
+    }
+
+    private getTextForHeader(element: HTMLElement): string {
+        const idText = `id="${this.getAttribute(element, 'id')}"`;
+        const headersAttr = this.getAttribute(element, 'headers');
+        const headersText = headersAttr === null ? null : `headers="${headersAttr}"`;
+
+        return ['th', idText, headersText].join(' ');
```

Comment:
This will show `th id="foo"` for `<div role="columnheader" id="foo">`; is that intended?

Reply:
My reasoning for this was that the text on the visual helper is already a bit cramped, and displaying "columnheader" on top of the table cells would take up more space and make it harder to read if the labels overlap. @LiLoDavis do you have any suggestions for whether we should be displaying something different for elements with roles (maybe something like the different capitalization we use for the headings visual helper)?

Reply:
Since we won't be including elements with columnheader/rowheader/gridcell roles (as per the thread below), this question is no longer relevant